File: app/src/main/java/com/rg/ireminders/db/utils/TaskUtils.java
package com.rg.ireminders.db.utils;

import android.content.Context;
import com.rg.ireminders.db.entities.TaskItem;
import com.rg.ireminders.db.entities.TaskList;
import com.rg.ireminders.db.utils.impl.TaskUtilsImpl;
import java.util.List;

/**
 * Created by rustamgaifullin on 2/7/15.
 */
public interface TaskUtils {
  List<TaskList> getTaskList();

  List<TaskItem> getTaskByTaskListId(Long taskListId);

  boolean insertTask(String taskName, Long taskListId);

  void updateTask(Long id, Long listId, String taskName);

  void changeTaskStatus(Long id, Long listId, boolean completed);

  void addReminder(Long id, Long listId, Long dueDate);

  void removeReminder(Long id, Long listId);

  int deleteCompleted(Long listId);

  public static class Factory {
    private static Factory instance = new Factory();

    public static TaskUtilsImpl get(Context context) {
      return instance.newForContext(context);
    }

    public static Factory getInstance() {
      return instance;
    }

    public static void overrideInstance(Factory factory) {
      instance = factory;
    }

    protected TaskUtilsImpl newForContext(Context context) {
      return new TaskUtilsImpl(context.getContentResolver());
    }
  }
}

File: app/src/main/java/com/rg/ireminders/db/utils/TaskUtils.kt
package com.rg.ireminders.db.utils

import android.content.Context
import com.rg.ireminders.db.entities.TaskItem
import com.rg.ireminders.db.entities.TaskList
import com.rg.ireminders.db.utils.impl.TaskUtilsImpl

/**
 * Created by rustamgaifullin on 2/7/15.
 */
interface TaskUtils {
    val taskList: List<TaskList>

    fun getTaskByTaskListId(taskListId: Long?): List<TaskItem>

    fun insertTask(taskName: String, taskListId: Long?): Boolean

    fun updateTask(id: Long?, listId: Long?, taskName: String)

    fun changeTaskStatus(id: Long?, listId: Long?, completed: Boolean)

    fun addReminder(id: Long?, listId: Long?, dueDate: Long?)

    fun removeReminder(id: Long?, listId: Long?)

    fun deleteCompleted(listId: Long?): Int

    val scheduledTasks: List<TaskItem>

    class Factory {

        protected fun newForContext(context: Context): TaskUtilsImpl {
            return TaskUtilsImpl(context.contentResolver)
        }

        companion object {
            var instance = Factory()
                private set

            operator fun get(context: Context): TaskUtilsImpl {
                return instance.newForContext(context)
            }

            fun overrideInstance(factory: Factory) {
                instance = factory
            }
        }
    }
}
